### PR TITLE
i#3544 RV64: Port sample 'cbr' to RISCV64

### PR DESCRIPTION
**Pull request by chenhy0106**
1. Conditional branch instruction of RISC-V 'C' extension may not reach after adding clean call. So like X86, we add support to detect and convert compressed cbr to longer version.
2. For RISCV64, when a cbr instruction uses the stolen reg, we need to replace it with a scratch reg. 

Now cbr works for RISCV64. Still does not work for ARM32/ARM64 because of some bugs unrelated to cbr sample.

Issue: #3544

### CODE REVIEW

**Review by ksco — COMMENTED** [review]
Looks like AArch64 CI is failing, and the cbr test is not enabled on RISC-V CI.

**Review by ksco — APPROVED** [review]

**Comment by derekbruening** [comment]
(May not get to this for another day)

File: api/samples/CMakeLists.txt [hunk 1]
```diff
@@ -241,8 +241,10 @@ add_sample_client(memtrace_simple "memtrace_simple.c;utils.c" "drmgr;drreg;druti
 add_sample_client(memval_simple   "memval_simple.c;utils.c"   "drmgr;drreg;drutil;drx")
 add_sample_client(instrace_simple "instrace_simple.c;utils.c" "drmgr;drreg;drx")
 add_sample_client(opcode_count    "opcode_count.cpp"    "drmgr;drreg;drx;droption")
+if (X86 OR RISCV64)
+add_sample_client(cbr         "cbr.c"           "drmgr")
```
Comment (derekbruening) [inline api/samples/CMakeLists.txt:245]: style: Indent 2 spaces

File: api/samples/CMakeLists.txt [hunk 2]
```diff
@@ -241,8 +241,10 @@ add_sample_client(memtrace_simple "memtrace_simple.c;utils.c" "drmgr;drreg;druti
 add_sample_client(memval_simple   "memval_simple.c;utils.c"   "drmgr;drreg;drutil;drx")
 add_sample_client(instrace_simple "instrace_simple.c;utils.c" "drmgr;drreg;drx")
 add_sample_client(opcode_count    "opcode_count.cpp"    "drmgr;drreg;drx;droption")
+if (X86 OR RISCV64)
```
Comment (derekbruening) [inline api/samples/CMakeLists.txt:244]: Please maintain the comment so it's clear we want to port this: `# TODO i#1551, #1569: Port to AArchXX.`

File: api/samples/cbr.c [hunk 3]
```diff
@@ -334,14 +338,46 @@ event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst
         app_pc fall = (app_pc)decode_next_pc(drcontext, (byte *)src);
         app_pc targ = instr_get_branch_target_pc(instr);
 
+#if defined(RISCV64)
+        /* For RISCV64, if instr uses the stolen regs, we have to replace it
```
Comment (derekbruening) [inline api/samples/cbr.c:342]: grammar: s/regs/reg/

File: core/lib/instrument.c [hunk 4]
```diff
@@ -7451,6 +7451,27 @@ dr_insert_it_instrs(void *drcontext, instrlist_t *ilist)
...
+    return false;
+#else
+    CLIENT_ASSERT(reg_is_pointer_sized(reg),
+                  "dr_insert_get_tp_reg_value: reg has wrong size\n");
+    CLIENT_ASSERT(reg != DR_REG_TP,
+                  "dr_insert_get_tp_reg_value: reg should not be tp itself\n");
+    instrlist_meta_preinsert(
+        ilist, instr,
+        instr_create_restore_from_tls(drcontext, reg,
+                                      os_get_app_tls_base_offset(TLS_REG_LIB)));
```
Comment (derekbruening) [inline core/lib/instrument.c:7469]: This seems to be the equivalent of dr_insert_get_seg_base().  Should we rename dr_insert_get_tp_reg_value to dr_insert_get_app_tls and have it call dr_insert_get_seg_base on x86, so we have a more generic routine?

**Review by derekbruening — APPROVED** [review]

**Comment by ksco** [comment]
LGTM, merging.

**Comment by ksco** [comment]
Hi @derekbruening, @chenhy0106 is an intern in our lab, and he would like to submit more PRs to DynamoRIO on RV64 support. Can we have him in the committer team so he can merge PRs himself in the future?

**Comment by derekbruening** [comment]
> Hi @derekbruening, @chenhy0106 is an intern in our lab, and he would like to submit more PRs to DynamoRIO on RV64 support. Can we have him in the committer team so he can merge PRs himself in the future?

Done.

**Pull request merged**